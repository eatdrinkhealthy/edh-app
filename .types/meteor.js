declare module "meteor/meteor" {

  declare type IMeteorSettings = {
    public: {
      googleMapsApiKey: string,
    },
    foursquare: {
      client_id: string,
      client_secret: string,
    }
  };

  declare export class IMeteorError {
    constructor(error: string, reason?: string, details?: string): IMeteorError;

    error: string;
    reason?: string;
    details?: string;
  }

  declare export var Meteor: {
    settings: IMeteorSettings,
    startup(): void,
    Error: IMeteorError,
    isClient: boolean,
    isServer: boolean,
  }
}
