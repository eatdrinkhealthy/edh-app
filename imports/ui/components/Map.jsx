// @flow
import React, {
  PureComponent,
} from "react";
import GoogleMap from "google-map-react";
import Marker from "./Marker";

// eslint-disable-next-line no-duplicate-imports, import/first
import type { ILatLng, IGoogleMapDisplay } from "google-map-react";
import type { IVenue } from "../../state/reducers/searchResultsReducers";

function createMapOptions() {
  return {
    fullscreenControl: false,
  };
}

export type IViewArea = {
  top: number,
  right: number,
  bottom: number,
  left: number,
};

export default class Map extends PureComponent {
  props: {
    center?: ILatLng,
    zoom?: number,
    googleMapsApiKey: string,
    venues: Array<IVenue>, // TODO can't this be optional with default? when so, produces flow error
    selectedVenueId: ?string,
    setSelectedVenueHandler: (venueId: ?string) => void,
    onMapChange?: (mapChange: ?IGoogleMapDisplay) => void,
  };

  static defaultProps = {
    center: { lat: 32.789008, lng: -79.932115 },
    zoom: 15,
    venues: [],
    onMapChange: (mapChange) => {}, // eslint-disable-line no-unused-vars
  }

  handleOnClick = () => {
    this.props.setSelectedVenueHandler(null);
  }

  mapHolder: HTMLDivElement;

  setMapHolderRef = (div: HTMLDivElement) => {
    this.mapHolder = div;
  }

  getHintViewArea = (): IViewArea => {
    const mhRect = this.mapHolder.getBoundingClientRect();

    const VIEW_AREA_HEADING_HEIGHT = 60;
    const VIEW_AREA_PADDING = 10;

    return {
      top: mhRect.top + VIEW_AREA_HEADING_HEIGHT + VIEW_AREA_PADDING,
      right: mhRect.right - VIEW_AREA_PADDING,
      bottom: mhRect.bottom - VIEW_AREA_PADDING,
      left: mhRect.left + VIEW_AREA_PADDING,
    };
  }

  render() {  // eslint-disable-line flowtype/require-return-type
    return (
      <div
        className="map-holder"
        ref={this.setMapHolderRef}
      >
        <GoogleMap
          bootstrapURLKeys={{ key: this.props.googleMapsApiKey }}
          defaultCenter={this.props.center}
          defaultZoom={this.props.zoom}
          onClick={this.handleOnClick}
          options={createMapOptions}
          onChange={this.props.onMapChange}
        >
          {this.props.venues.map((venue: IVenue): React$Element<*> => (
            <Marker
              key={venue.id}
              venue={venue}
              lat={venue.location.lat}
              lng={venue.location.lng}
              setSelectedVenueHandler={this.props.setSelectedVenueHandler}
              selected={venue.id === this.props.selectedVenueId}
              getHintViewArea={this.getHintViewArea}
            />))
          }
        </GoogleMap>
      </div>
    );
  }
}
