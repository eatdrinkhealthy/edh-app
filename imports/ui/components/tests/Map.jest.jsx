// @flow
/* eslint-env jest */
/* eslint-disable func-names, prefer-arrow-callback, no-unused-expressions */
/* eslint-disable import/no-extraneous-dependencies */

import React from "react";
import { Meteor } from "meteor/meteor";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Map from "../Map";
import sampleVenues from "../../../state/stores/tests/sampleVenueData";

describe("<Map />", function () {
  it("matches render snapshot", function () {
    // TODO - to capture more snapshot detail, use mount or react-test-renderer (BOTH FAIL HERE)
    const wrapper = shallow(<Map
      googleMapsApiKey={Meteor.settings.public.googleMapsApiKey}
      setSelectedVenueHandler={() => {}}
      selectedVenueId={null}
      center={{ lat: 32.789008, lng: -79.932115 }}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("matches render snapshot - with userPosition pin", function () {
    const wrapper = shallow(<Map
      googleMapsApiKey={Meteor.settings.public.googleMapsApiKey}
      setSelectedVenueHandler={() => {}}
      selectedVenueId={null}
      userPosition={{ lat: 32.789008, lng: -79.932115 }}
      center={{ lat: 32.789008, lng: -79.932115 }}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("Map markers", function () {
    const testVenues = [
      sampleVenues[0],
      sampleVenues[1],
      sampleVenues[2],
    ];

    // eslint-disable-next-line flowtype/require-return-type, flowtype/require-parameter-type
    const selectedPropTrue = node => node.props().selected === true;

    it("should display provided search results as markers", function () {
      const wrapper = shallow(<Map
        googleMapsApiKey={Meteor.settings.public.googleMapsApiKey}
        venues={testVenues}
        setSelectedVenueHandler={() => {}}
        selectedVenueId={null}
        center={{ lat: 32.789008, lng: -79.932115 }}
      />);

      expect(wrapper.find("Marker").length).toBe(3);
    });

    it("should have no 'selected' markers when selectedVenueId provided null", function () {
      const wrapper = shallow(<Map
        googleMapsApiKey={Meteor.settings.public.googleMapsApiKey}
        venues={testVenues}
        setSelectedVenueHandler={() => {}}
        selectedVenueId={null}
        center={{ lat: 32.789008, lng: -79.932115 }}
      />);

      expect(wrapper.find("Marker").findWhere(selectedPropTrue).length).toBe(0);
    });

    it("should have one 'selected' marker when provided a matching selectedVenueId", function () {
      const wrapper = shallow(<Map
        googleMapsApiKey={Meteor.settings.public.googleMapsApiKey}
        venues={testVenues}
        setSelectedVenueHandler={() => {}}
        selectedVenueId="3"
        center={{ lat: 32.789008, lng: -79.932115 }}
      />);

      expect(wrapper.find("Marker").findWhere(selectedPropTrue).length).toBe(1);
    });

    it("should clear a selected marker when the map is clicked", function () {
      const selectVenue = jest.fn();
      const wrapper = shallow(<Map
        googleMapsApiKey={Meteor.settings.public.googleMapsApiKey}
        venues={testVenues}
        setSelectedVenueHandler={selectVenue}
        selectedVenueId="3"
        center={{ lat: 32.789008, lng: -79.932115 }}
      />);

      wrapper.find("GoogleMap").simulate("click");
      expect(selectVenue).toHaveBeenCalledWith(null);
    });
  });
});
