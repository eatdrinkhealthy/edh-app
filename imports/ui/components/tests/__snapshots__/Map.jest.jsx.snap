// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Map /> matches render snapshot 1`] = `
<div
  className="map-holder"
>
  <GoogleMap
    bootstrapURLKeys={
      Object {
        "key": "TestAPIKeyValue",
      }
    }
    center={
      Object {
        "lat": 32.789008,
        "lng": -79.932115,
      }
    }
    debounced={true}
    distanceToMouse={[Function]}
    googleMapLoader={[Function]}
    hoverDistance={30}
    layerTypes={Array []}
    onChange={[Function]}
    onClick={[Function]}
    options={[Function]}
    style={
      Object {
        "height": "100%",
        "margin": 0,
        "padding": 0,
        "position": "relative",
        "width": "100%",
      }
    }
    yesIWantToUseGoogleMapApiInternals={false}
    zoom={15}
  />
</div>
`;
